stdlib/fileinput.pyi

```diff
@@ -65,36 +162,131 @@ def isstdin() -> bool: ...
+            inplace: bool = ...,
+            backup: str = ...,
+            *,
+            mode: Literal["rb"],
+            openhook: Callable[[StrOrBytesPath, str], _HasReadlineAndFileno[bytes]] | None = ...,
+            encoding: None = ...,
+            errors: None = ...,
+        ) -> None: ...
+        @overload
+        def __init__(
+            self: FileInput[Any],
+            files: StrOrBytesPath | Iterable[StrOrBytesPath] | None = ...,
+            inplace: bool = ...,
+            backup: str = ...,
+            *,
+            mode: str,
+            openhook: Callable[[StrOrBytesPath, str], _HasReadlineAndFileno[Any]] | None = ...,
+            encoding: str | None = ...,
+            errors: str | None = ...,
+        ) -> None: ...
+
     elif sys.version_info >= (3, 8):
+        # bufsize is dropped and mode and openhook become keyword-only
+        @overload
+        def __init__(
+            self: FileInput[str],
+            files: StrOrBytesPath | Iterable[StrOrBytesPath] | None = ...,
+            inplace: bool = ...,
+            backup: str = ...,
+            *,
+            mode: _TextMode = ...,
+            openhook: Callable[[StrOrBytesPath, str], _HasReadlineAndFileno[str]] | None = ...,
+        ) -> None: ...
+        @overload
+        def __init__(
+            self: FileInput[bytes],
+            files: StrOrBytesPath | Iterable[StrOrBytesPath] | None = ...,
+            inplace: bool = ...,
+            backup: str = ...,
+            *,
+            mode: Literal["rb"],
+            openhook: Callable[[StrOrBytesPath, str], _HasReadlineAndFileno[bytes]] | None = ...,
+        ) -> None: ...
+        @overload
         def __init__(
-            self,
-            files: None | StrOrBytesPath | Iterable[StrOrBytesPath] = ...,
+            self: FileInput[Any],
+            files: StrOrBytesPath | Iterable[StrOrBytesPath] | None = ...,
             inplace: bool = ...,
             backup: str = ...,
             *,
-            mode: str = ...,
-            openhook: Callable[[StrOrBytesPath, str], IO[AnyStr]] = ...,
+            mode: str,
+            openhook: Callable[[StrOrBytesPath, str], _HasReadlineAndFileno[Any]] | None = ...,
         ) -> None: ...
+
     else:
+        @overload
         def __init__(
-            self,
-            files: None | StrOrBytesPath | Iterable[StrOrBytesPath] = ...,
+            self: FileInput[str],
+            files: StrOrBytesPath | Iterable[StrOrBytesPath] | None = ...,
             inplace: bool = ...,
             backup: str = ...,
             bufsize: int = ...,
-            mode: str = ...,
-            openhook: Callable[[StrOrBytesPath, str], IO[AnyStr]] = ...,
+            mode: _TextMode = ...,
+            openhook: Callable[[StrOrBytesPath, str], _HasReadlineAndFileno[str]] | None = ...,
+        ) -> None: ...
+        # Because mode isn't keyword-only here yet, we need two overloads each for
+        # the bytes case and the fallback case.
```

Comment:
Sebastian opened a conversation a while ago about allowing arguments without defaults to follow arguments with defaults, which would make this easier, but it didn't go anywhere.

Reply:
There was some pushback, because that's mostly useful for typing and considered bad API design otherwise, but honestly more from typing-critical developers, if I remember correctly.

Reply:
> An alternative could be to add a special marker like `typing.NO_DEFAULT`, which we'd use as the default.

Oh, I like that quite a lot, actually. It would mean we wouldn't have to change Python's grammar at all. It does also feel like the kind of thing that would disproportionately benefit typeshed, though.

Reply:
It should be possible to write a script that automatically applies `NO_DEFAULT` in typeshed wherever possible. I would be interested to see the results,